docs/proposals/gpu-support.md

```diff
@@ -0,0 +1,305 @@
+available GPU resources. It won't return multiple nodes, because the API does
+not inform the extender about the node that is eventually picked. Without the
+node's identity, resources can't be reserved permanently. It is possible to
+obtain the information indirectly by placing a watch on the pod, but there might
+be race conditions between delivery of the notification and the next incoming
+extender request. Resources are thus committed right away to the pod on that
+specific node (this can be a temporary lease, in case the pod can't schedule
+there for other reasons). Because of that, the GPU extender should be the last
+in the scheduler policy configuration. Actually, a watch will still be created,
+but, instead, it will be used to return the resources used by the pod when it
+terminates.
+
+Nodes with available resources are handed out in a circular fashion in order to
+
+- approximate round robin scheduling
+- increase the chances of forward progress when the same pod is passed to the
+  extender multiple times (e.g. because of other constraints not being met)
+
+### The runtime
+
+Once we know where to run the container, it's time to set up its environment. At
+a minimum, we'll need to map the host device(s) into the container. Because each
+manufacturer exposes different device nodes (`/dev/ati/card0`, `/dev/nvidia0`,
+but also the required `/dev/nvidiactl` and `/dev/nvidia-uvm`), some of the logic
+needs to be hardware-specific, mapping from a logical device to a list of device
+nodes necessary for software to talk to it.
+
+Support binaries and libraries are often versioned along with the kernel module,
+so there should be further hooks to project those under `/bin` and some kind of
+`/lib` before the application is started. This can be done for Docker with the
+use of a versioned [Docker
+volume](https://docs.docker.com/engine/userguide/containers/dockervolumes/). The
+alternative would be for each application to bundle everything in their image.
+
+
+#### NVIDIA support
+
+The first implementation and proof of concept will be for NVIDIA devices, the
+most common setup, also with more extensive infrastructure, namely the
+[nvidia-docker plugin](https://github.com/NVIDIA/nvidia-docker). The plugin is
+an officially supported solution, thus avoiding a lot of new low level code, as
+it takes care of functionality such as:
+
+- creating a Docker volume with binaries such as `nvidia-smi` and shared
+libraries
+- providing HTTP endpoints that monitoring tools can use to collect GPU metrics
+- abstracting details such as `/dev` entry names for each device, as well as
+control ones like `nvidiactl`
+
+The `nvidia-docker` wrapper also verifies that the CUDA version required by a
+given image is supported by the host drivers, through inspection of well-known
+image labels, if present. We should try to provide equivalent checks.
+
+This is current sample output from the plugin, wrapped for readability:
+
+    $ curl -s localhost:3476/docker/cli
```

Comment:
The plugin runs as a standalone daemon, listening to HTTP requests. One of the endpoints returns flags to add to Docker command lines. By default, it assumes you want to run on device 0, but you can change the URL to provide a list of device numbers instead.